harness/determined/pytorch/_pytorch_context.py

```diff
@@ -572,7 +580,7 @@ def backward(
                         gradient=gradient, retain_graph=retain_graph, create_graph=create_graph
                     )
 
-                if self.distributed.size > 1 and self._should_communicate_and_update():
+                if self.distributed.size > 1 and self._should_communicate_and_update() and self._distributed_backend.use_horovod():
                     # When we exit out of this context manager, we need to finish
                     # communicating gradient updates before they are unscaled.
```

Comment:
this looks to me like a bug:

- Losses are unscaled when we exit the apex.amp.scale_loss() context manager.
- Losses need to be synchronized before they are unscaled or the scalers can get out-of-sync, causing training to diverge
- In DDP we won't be synchronizing gradients inside the scale_loss context manager, and therefore we have a bug.

Please take a look, and correct me if I'm wrong.

Reply:
hm... is it that DDP losses are synchronized during the call to .backward?  That would be worth a comment if it is the case.

Reply:
yep. DDP receives a signal at the backward() pass to synchronize gradients. i will update the README on all of this.